### PR TITLE
Conditional mysql compilation and other tweaks

### PR DESCRIPTION
**Pull request by wongsyrone**
- conditional mysql
- don't pollute CXX flags
- modify flags when not defined

### CODE REVIEW

**Review by GreaterFire — APPROVED** [review]

**Comment by wongsyrone** [comment]
```
x64 BUILD COMMANDS

mkdir build
pushd build
..\..\cmake-3.12.1-win64-x64\bin\cmake.exe -G "Visual Studio 15 Win64" -DENABLE_MYSQL=OFF -DBOOST_ROOT=D:\github\boost-install -DBoost_NO_SYSTEM_PATHS=ON -DBoost_USE_DEBUG_LIBS=OFF -DBoost_USE_RELEASE_LIBS=ON -DBoost_USE_STATIC_LIBS=ON -DBoost_USE_STATIC_RUNTIME=ON -DOPENSSL_ROOT_DIR=D:\github\openssl-1.1.0i\x64-build ..

if you are using `-DBoost_USE_STATIC_RUNTIME=ON` Change from /MD to /MT
a.k.a modify 'RuntimeLibrary' of 'Release|x64' in trojan.vcxproj to 'MultiThreaded' instead of 'MultiThreadedDLL'

..\..\cmake-3.12.1-win64-x64\bin\cmake.exe --build . --config Release

popd
```

**Comment by wongsyrone** [comment]
It seems newer version adds more libraries, e.g. boost 1.68.0
```
 Directory of D:\github\boost-install\lib

2018/08/17  14:17    <DIR>          .
2018/08/17  14:17    <DIR>          ..
2018/08/17  14:06             7,188 libboost_atomic-vc141-mt-s-x64-1_68.lib
2018/08/17  14:06           414,480 libboost_chrono-vc141-mt-s-x64-1_68.lib
2018/08/17  14:06           282,958 libboost_container-vc141-mt-s-x64-1_68.lib
2018/08/17  14:06            63,154 libboost_context-vc141-mt-s-x64-1_68.lib
2018/08/17  14:06         1,348,988 libboost_contract-vc141-mt-s-x64-1_68.lib
2018/08/17  14:07           444,326 libboost_coroutine-vc141-mt-s-x64-1_68.lib
2018/08/17  09:59           800,058 libboost_date_time-vc141-mt-s-x64-1_68.lib
2018/08/17  14:07             1,564 libboost_exception-vc141-mt-s-x64-1_68.lib
2018/08/17  14:08         2,710,352 libboost_fiber-vc141-mt-s-x64-1_68.lib
2018/08/17  14:07         1,217,222 libboost_filesystem-vc141-mt-s-x64-1_68.lib
2018/08/17  14:09         7,756,358 libboost_graph-vc141-mt-s-x64-1_68.lib
2018/08/17  14:09           624,756 libboost_iostreams-vc141-mt-s-x64-1_68.lib
2018/08/17  14:10        11,471,006 libboost_locale-vc141-mt-s-x64-1_68.lib
2018/08/17  14:12        22,902,348 libboost_log-vc141-mt-s-x64-1_68.lib
2018/08/17  14:14        26,672,190 libboost_log_setup-vc141-mt-s-x64-1_68.lib
2018/08/17  14:15           675,928 libboost_prg_exec_monitor-vc141-mt-s-x64-1_68.lib
2018/08/17  14:14        10,215,420 libboost_program_options-vc141-mt-s-x64-1_68.lib
2018/08/17  14:14           167,096 libboost_random-vc141-mt-s-x64-1_68.lib
2018/08/17  14:08        16,879,648 libboost_regex-vc141-mt-s-x64-1_68.lib
2018/08/17  14:14        13,013,442 libboost_serialization-vc141-mt-s-x64-1_68.lib
2018/08/17  14:15         1,230,768 libboost_signals-vc141-mt-s-x64-1_68.lib
2018/08/17  14:15            16,352 libboost_stacktrace_noop-vc141-mt-s-x64-1_68.lib
2018/08/17  14:15           135,204 libboost_stacktrace_windbg-vc141-mt-s-x64-1_68.lib
2018/08/17  14:15           138,610 libboost_stacktrace_windbg_cached-vc141-mt-s-x64-1_68.lib
2018/08/17  14:06           107,828 libboost_system-vc141-mt-s-x64-1_68.lib
2018/08/17  14:15        19,609,674 libboost_test_exec_monitor-vc141-mt-s-x64-1_68.lib
2018/08/17  14:07         1,046,110 libboost_thread-vc141-mt-s-x64-1_68.lib
2018/08/17  14:15           478,600 libboost_timer-vc141-mt-s-x64-1_68.lib
2018/08/17  14:16           543,044 libboost_type_erasure-vc141-mt-s-x64-1_68.lib
2018/08/17  14:15        19,422,908 libboost_unit_test_framework-vc141-mt-s-x64-1_68.lib
2018/08/17  14:17        45,209,692 libboost_wave-vc141-mt-s-x64-1_68.lib
2018/08/17  14:15        10,601,048 libboost_wserialization-vc141-mt-s-x64-1_68.lib
```

**Comment by wongsyrone** [comment]
Boost 1.68.0 build commands

check what libraries need to be really compiled

```
b2 --show-libraries
```

use this to compile
```
b2 --prefix=D:\github\boost-install architecture=x86 address-model=64 toolset=msvc-14.1 variant=release link=static threading=multi runtime-link=static --without-mpi --without-math --without-python install
```

use this to clean builds

```
b2 --clean-all
```

**Review by GreaterFire — APPROVED** [review]

File: CMakeLists.txt [hunk 1]
```diff
@@ -5,29 +5,27 @@ if(NOT CMAKE_BUILD_TYPE)
...
+    src/socks5address.cpp
+    src/ssldefaults.cpp
+    src/trojanrequest.cpp
+    src/udppacket.cpp
+    src/version.cpp)
 find_package(Threads REQUIRED)
 target_link_libraries(trojan ${CMAKE_THREAD_LIBS_INIT})
-# COMPONENTS is used to generate linkage flags, e.g.: MSVC
-# It should match what we are using
-find_package(Boost 1.54.0 REQUIRED COMPONENTS system date_time)
```
Comment (wongsyrone) [inline CMakeLists.txt]: Why？It's required to generate msbuild project

File: CMakeLists.txt [hunk 2]
```diff
@@ -38,7 +38,7 @@ if(ENABLE_MYSQL)
     include_directories(${MYSQL_INCLUDE_DIR})
     target_link_libraries(trojan ${MYSQL_LIBRARIES})
     add_definitions(-DENABLE_MYSQL)
-endif(ENABLE_MYSQL)
```
Comment (wongsyrone) [inline CMakeLists.txt]: 这样写是为了确定对应关系，和你写ifdef的时候是一样的

File: CMakeLists.txt (hunk 1 above)
Comment (GreaterFire) [reply CMakeLists.txt]: Fixed in https://github.com/trojan-gfw/trojan/pull/40/commits/4ecc2ac272f32c624720c1c135289804a770741f#diff-af3b638bc2a3e6c650974192a53c7291R31.

File: CMakeLists.txt (hunk 2 above)
Comment (GreaterFire) [reply CMakeLists.txt]: Unlike ifdef, with indentation, it's easier to find the relationship.

**Review by GreaterFire — APPROVED** [review]

**Pull request merged**